services/addons/images/firmware_manager/yunex_upgrader.py

```diff
@@ -0,0 +1,117 @@
+import upgrader
+import json
+import logging
+import os
+import subprocess
+import sys
+import tarfile
+import time
+
+class YunexUpgrader( upgrader.UpgraderAbstractClass ):
+  def __init__(self, upgrade_info):
+    super().__init__(upgrade_info)
+
+  def run_xfer_upgrade(self, file_name):
+    xfer_command = ["java", "-jar", f"/home/tools/xfer_yunex.jar", "-upload", file_name, f"{self.rsu_ip}:3600"]
+    proc = subprocess.run(xfer_command, capture_output=True)
+    code, stdout, stderr = proc.returncode, proc.stdout, proc.stderr
+
+    # If the command ends with a non-successful status code, return -1
+    if code != 0:
+      logging.error("Firmware not successful: " + stderr.decode("utf-8"))
+      return -1
+
+    output_lines = stdout.decode("utf-8").split('\n')[:-1]
+    # If the command ends with a successful status code but the logs don't contain the expected line, return -1
+    if 'TEXT: {"success":{"upload":"Processing OK. Rebooting now ..."}}' not in output_lines:
+      logging.error("Firmware not successful: " + stderr.decode("utf-8"))
+      return -1
+
+    # If everything goes as expected, the XFER upgrade was complete
+    return 0

```

Comment:
Might be good to have some debug logs that outputs if the upgrade is successful.

Reply:
I took those out since it is verbose if it isn't